machine/machine_core/machine.py

```diff
@@ -357,3 +358,93 @@ def curl(self, *args: str, headers: Mapping[str, str] = {}) -> str:
         return self.execute(cmd + list(args))
+
+    def clone_container(self) -> "NspawnMachine":
+        """Clone the host OS/file system into a container
+
+        This uses nspawn to boot the host OS in a container running on the same machine, with
+        an ephemeral file system overlay. This is useful for tests which need multiple
+        independent machines in terms of network/file system/package separation, but no
+        kernel modules or hardware access.
+        """
+        return NspawnMachine(self, self.verbose)
+
+
+class NspawnMachine(Machine):
+    def __init__(
+        self,
+        parent: Machine,
+        verbose: bool = False,
+    ):
+        self.parent = parent
+
+        # find a free container name
+        machines = []
+        for line in parent.execute("machinectl list --no-legend").strip().splitlines():
+            machines.append(line.split()[0])
+
+        # FIXME: the SSHConnection master gets confused with more than one container,
+        # so disallow this for now until this gets fixed
+        if machines:
+            raise RuntimeError("Only one container is supported at the moment")
+
+        for i in range(10):
+            label = f"nspawn{i}"
+            if label not in machines:
+                break
+        else:
+            raise RuntimeError("did not find a free nspawn machine label")
+
+        self.nspawn_label = label
+
+        # networkd auto-configures host and guest veth
+        parent.execute("systemctl enable --now systemd-networkd")
+
+        # we already have to start the container here (as opposed to start()) to figure out its address
+        # ideally we could just use --ephemeral here, but it's way too inefficient on non-btrfs
+        overlays = [f"--overlay {d}:$O/{d}:{d}" for d in ["/etc", "/var", "/usr", "/home", "/root"]]
+        self.nspawn_pid = parent.spawn(
+            f"O=/var/tmp/overlay-{self.nspawn_label}; "
+            "mkdir -p $O/etc $O/var $O/usr $O/home $O/root; "
+            f"systemd-nspawn --machine={label} --network-veth --directory=/ --boot --volatile=yes " +
+            " ".join(overlays),
+            label)
+
+        # wait until the network initialized, and get the container address
+        # this is horrible.. is there a better way? `machinectl show` doesn't have the address
+        address = parent.execute(f"""while true; do
+                                        A=$(machinectl list | sed -rn '/^{label}.* [0-9.]+$/ {{ s/^.* //; p }}');
+                                        [ -z "$A" ] || break;
+                                        sleep 0.5;
+                                    done;
+                                    echo $A""", timeout=20).strip()
+
+        # wait until SSH is up, so that the ProxyCommand doesn't get stuck
+        parent.execute(f"until timeout 1 nc -z {address} 22; do sleep 1; done")
+
+        super().__init__(
+            address, parent.image, verbose, parent.label + "-" + label,
```

Comment:
Looks like `self.verbose` should be initialized earlier, in case the code above throws an error.